### PR TITLE
MolSSI Driver Interface support

### PR DESCRIPTION
**Pull request by taylor-a-barnes**
This PR enables QCEngine to function as a engine within the context of the MolSSI Driver Interface ([MDI](https://molssi.github.io/MDI_Library/html/index.html)).  MDI supports interoperability within a driver-engine paradigm, in which a driver code controls the high-level operations of one or more engine codes, orchestrating complex calculations like QM/MM or advanced sampling.  This PR allows MDI to interface with QCEngine, controlling production codes using QCEngine as an intermediary.

The following is an example script for launching QCEngine as an MDI engine:

```
import qcengine as qcng
import qcelemental as qcel

mol = qcel.models.Molecule.from_data("""                                                                                                     
O 4.92718 4.99809 1.10308                                                                                                                    
H 5.80905 5.31488 0.90767                                                                                                                    
H 4.87602 4.14701 0.667997                                                                                                                   
O 0.0880888 4.00127 0.245364                                                                                                                 
H 9.67931 3.21326 0.714742                                                                                                                   
H 0.329457 3.6894 9.23817                                                                                                                    
O 8.38355 2.57034 3.9726                                                                                                                     
H 7.44821 2.36788 3.95264                                                                                                                    
H 8.7019 2.15517 4.77417                                                                                                                     
O 7.00057 7.16334 4.02289                                                                                                                    
H 7.72912 7.3014 3.41759                                                                                                                     
H 7.21545 6.34743 4.47493                                                                                                                    
O 4.30406 0.919395 2.3125                                                                                                                    
H 4.6805 1.40121 1.57604                                                                                                                     
H 4.49956 9.86655 2.12387                                                                                                                    
O 3.14136 6.90125 0.70106                                                                                                                    
H 2.42226 6.62251 1.26801                                                                                                                    
H 3.71511 6.13692 0.647808                                                                                                                   
O 7.16357 4.36105 7.68273                                                                                                                    
H 7.35485 3.47729 7.99679                                                                                                                    
H 7.40743 4.93254 8.41087                                                                                                                    
O 9.36571 1.39718 6.75649                                                                                                                    
H 9.80707 2.14953 7.15071                                                                                                                    
H 8.72182 1.12662 7.41103                                                                                                                    
""")

o = qcng.MDIServer(mdi_options = "-role ENGINE -name QM -method TCP -port 8021 -hostname localhost",
                   program = "psi4",
                   molecule = mol,
                   model = {"method": "SCF", "basis": "sto-3g"},
                   keywords = {"scf_type": "df"})
o.start()
```

### CODE REVIEW

**Comment by taylor-a-barnes** [comment]
@dgasmith Thanks for these comments!  I believe this is now ready to go.

File: qcengine/mdi_server.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+        """
+        charge = self.molecule.molecular_charge
+        MDI_Send(charge, 1, MDI_DOUBLE, self.comm)
+        return charge
+
+    # Respond to the >TOTCHARGE command
+    def recv_total_charge(self, charge=None):
+        """ Receive the total system charge through MDI
+
+        Arguments:
```
Comment (dgasmith) [inline qcengine/mdi_server.py]: Could you use Napoleon style docs here? You can see this style in the associated files.

File: qcengine/mdi_server.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+        """ Send the total system charge through MDI
+
+        :returns: *charge* Total charge of the system
+        """
+        charge = self.molecule.molecular_charge
+        MDI_Send(charge, 1, MDI_DOUBLE, self.comm)
+        return charge
+
+    # Respond to the >TOTCHARGE command
+    def recv_total_charge(self, charge=None):
```
Comment (dgasmith) [inline qcengine/mdi_server.py]: Would it be possible to add types here as well? ` def recv_total_charge(self, charge:Optional[float]=None) -> None:

**Comment by taylor-a-barnes** [comment]
@dgasmith Thanks for reviewing this, and sorry for the slow response.  I think I'll be able to address the last few items tomorrow.

**Comment by taylor-a-barnes** [comment]
@dgasmith The latest commit should address your comments.  Thanks again!

**Comment by dgasmith** [comment]
Looks good, thanks!

**Pull request merged**